lightning-invoice/src/lib.rs

```diff
@@ -1648,7 +1648,7 @@ impl<'de> Deserialize<'de> for Invoice {
 	fn deserialize<D>(deserializer: D) -> Result<Invoice, D::Error> where D: Deserializer<'de> {
 		let bolt11 = String::deserialize(deserializer)?
 			.parse::<Invoice>()
-			.map_err(|e| D::Error::custom(format!("{:?}", e)))?;
+			.map_err(|e| D::Error::custom(alloc::format!("{:?}", e)))?;
```

Comment:
Just changing `format!` to `format_args!` would've been better since it may avoid allocation.

Reply:
Firstly, if you call `format!` you have two allocations - one for the string stored inside serde error, the other for the temporary string. So it should be always better to use `format_args!`.

Secondly while it's true that most if not all serializers in practice do store string, it's possible that someone somewhere decides to write a serializer that directly logs the errors instead of storing them just to minimize allocations here.

Reply:
What extra lifetime? `serde::de::Error::custom` takes any `T: Display`, there's not `'static` bound.

Reply:
It won't, `T: Display` means any `T`, including one containing references. I'd have to be `T: Display + 'static`.

Presumably most implementors internally call `to_string()` aon the argument and store the string. If you call it with a `String` it'll call `to_string()` on it making a clone because of taking shared reference.

Reply:
That's not a problem because `E::custom` doesn't store the argument inside (the signature prevents it). It either stores a `String` (possibly something exotic like `Arc<str>`) or logs it.

Anyway, no point in trying to convince you by arguing if I can make a [PR](https://github.com/lightningdevkit/rust-lightning/pull/2262) :)

Reply:
I see. I think casing (`custom` vs `Custom`) should've made it clear?